sdk/src/types/block/mana/parameters.rs

```diff
@@ -196,6 +196,37 @@ impl ProtocolParameters {
                 - (c >> mana_parameters.decay_factors_exponent())
         })
     }
+
+    pub fn slots_until_generated(
+        &self,
+        generation_amount: u64,
+        stored_mana: u64,
+        required_mana: u64,
+    ) -> Result<u32, ManaError> {
+        let mut num_slots = 0;
+        let mana_generated_per_epoch = self
+            .mana_parameters()
+            .generate_mana(generation_amount, self.slots_per_epoch());
+        let mut required_mana_remaining = required_mana;
+        loop {
+            // Get the minimum number of slots required to achieve the needed mana (i.e. not including decay)
+            num_slots +=
+                u32::try_from(1 + (required_mana_remaining * self.slots_per_epoch() as u64) / mana_generated_per_epoch)
```

Comment:
Probably missing an extra check for mana_generated_per_epoch

thread 'client::transaction_builder::outputs::insufficient_mana2' panicked at sdk/src/types/block/mana/parameters.rs:214:35:
attempt to divide by zero

```
#[test]
fn insufficient_mana2() {
    let protocol_parameters = iota_mainnet_protocol_parameters();

    let min_amount = BasicOutputBuilder::new_with_minimum_amount(protocol_parameters.storage_score_parameters())
        .with_mana(0)
        .add_unlock_condition(AddressUnlockCondition::new(
            Address::try_from_bech32(BECH32_ADDRESS_ED25519_0).unwrap(),
        ))
        .finish_output()
        .unwrap()
        .amount();

    let inputs = build_inputs(
        [
            (
                Basic {
                    amount: min_amount,
                    mana: 0,
                    address: Address::try_from_bech32(BECH32_ADDRESS_ED25519_0).unwrap(),
                    native_token: None,
                    sender: None,
                    sdruc: None,
                    timelock: None,
                    expiration: None,
                },
                None,
            ),
            (
                Basic {
                    amount: min_amount,
                    mana: 0,
                    address: Address::try_from_bech32(BECH32_ADDRESS_ED25519_0).unwrap(),
                    native_token: None,
                    sender: None,
                    sdruc: None,
                    timelock: None,
                    expiration: None,
                },
                None,
            ),
        ],
        Some(SLOT_INDEX),
    );

    let outputs = build_outputs([Basic {
        amount: 28200,
        mana: 10000,
        address: Address::try_from_bech32(BECH32_ADDRESS_ED25519_0).unwrap(),
        native_token: None,
        sender: None,
        sdruc: None,
        timelock: None,
        expiration: None,
    }]);

    let selected = TransactionBuilder::new(
        inputs.clone(),
        outputs.clone(),
        [Address::try_from_bech32(BECH32_ADDRESS_ED25519_0).unwrap()],
        SLOT_INDEX,
        SLOT_COMMITMENT_ID,
        protocol_parameters.clone(),
    )
    .finish()
    .unwrap_err();

    let TransactionBuilderError::InsufficientMana {
        found,
        required,
        slots_remaining,
    } = selected
    else {
        panic!("expected insufficient mana error")
    };
    assert_eq!(found, 0);
    assert_eq!(required, 10000);
}
```

Reply:
Ah, good point, thank you

Reply:
If generation_amount is very low and doesn't generate enough mana it still fails because `mana_generated_per_epoch` will still be 0, so probably needs the check for `mana_generated_per_epoch`(for generation_amount it's not needed?)
```Rust
        if mana_generated_per_epoch == 0 {
            return Err(ManaError::InsufficientGenerationAmount);
        }
```

Reply:
Hmm, fair point